### PR TITLE
[css-text-decor-4] Fix the expected serialization of text-decoration.

### PR DESCRIPTION
**Pull request by BorisChiou**
We add some test cases for `text-decoration-thickness` in `text-decoration`.
    
Conceptually the process of serializing a shorthand in getComputedStyle should be something like:
* Calling to_resolved_value() on all the values for the properties of the shorthand. This should turn the color as an rgb(a) thing, as colors are always serialized as resolved from getComputedStyle().
* Going from the resolved value al the way to the specified value. This makes the color still an rgba color, not currentColor.
* Serialize the same way we serialize specified shorthands.
    
So we always serizlize `text-decoration-color` because it is impossible to be `currentColor`. For other keywords, if they are initial values, we omit them.

### CODE REVIEW

**Comment by BorisChiou** [comment]
@ericwilligers This may need your review.

cc @emilio

File: css/css-text-decor/parsing/text-decoration-valid.html [hunk 1]
```diff
@@ -2,7 +2,8 @@
 <meta charset="utf-8">
 <title>CSS Text Decoration Test: Parsing text-decoration with valid values</title>
 <link rel="help" href="https://drafts.csswg.org/css-text-decor-3/#propdef-text-decoration">
-<meta name="assert" content="text-decoration supports the full grammar '<‘text-decoration-line’> || <‘text-decoration-style’> || <‘text-decoration-color’>'.">
+<link rel="help" href="https://drafts.csswg.org/css-text-decor-4/#text-decoration-width-property">
+<meta name="assert" content="text-decoration supports the full grammar '<‘text-decoration-line’> || <‘text-decoration-style’> || <‘text-decoration-color’> || <'text-decoration-thickness>'.">
```
Comment (ewilligers) [inline css/css-text-decor/parsing/text-decoration-valid.html]: text-decoration-width is not yet tested in the file.

Thus we should not yet mention css-text-decor-4 or text-decoration-width.

**Comment by ewilligers** [comment]
Are you suggesting the serialization of text-decoration from gCS should be different in the following two cases? Is there a precedent?

```
color: blue;
text-decoration-style: dotted;
text-decoration-color: currentColor;
```

```
color: blue;
text-decoration-style: dotted;
text-decoration-color: blue;
```

**Comment by BorisChiou** [comment]
> Are you suggesting the serialization of text-decoration from gCS should be different in the following two cases? Is there a precedent?

This is a good example. I didn't notice any precedent, at least now. Emilio suggests to me that it'd be better to always serialize `color` because its resolved value is never currentColor. For other keywords, if they are default values, we omit them. I will update this patch to let you review again. Thanks for the question and example.

File: css/css-text-decor/parsing/text-decoration-valid.html (hunk 1 above)
Comment (BorisChiou) [reply css/css-text-decor/parsing/text-decoration-valid.html]: I see. I will drop this.

Comment (BorisChiou) [reply css/css-text-decor/parsing/text-decoration-valid.html]: Actually we did. At least these three tests are related to text-decoration-thickness (which has been renamed from text-decoration-width). However, the property name is text-decoration, so ok I will drop this. Thanks. 
```
test_valid_value("text-decoration", "auto", "none");
test_valid_value("text-decoration", "from-font");
test_valid_value("text-decoration", "10px");
```

**Comment by BorisChiou** [comment]
Updated the description.

**Comment by ewilligers** [comment]
> Serialize the same way we serialize specified shorthands.

Are you saying that for the specified value, `none` and `solid` should be omitted whenever a color is specified, or are  `none` and `solid` always omitted, and `currentcolor` added if the result would otherwise be empty?

**Comment by BorisChiou** [comment]
Sorry for the confusing description. I reply these below.

> Are you saying that for the specified value, none and solid should be omitted whenever a color is specified

For specified shorthand, we omit none and solid if the color is not currentcolor. For example:
```
div.style.textDecoration = "none rgb(0, 0, 255)";
console.log(div.style.textDecoration); // output is "rgb(0, 0, 255)"

div.style.textDecoration = "solid rgb(0, 0, 255)";
console.log(div.style.textDecoration); // output is "rgb(0, 0, 255)"
```

> or are none and solid always omitted, and currentcolor added if the result would otherwise be empty? For example, how should a specified value of none solid be serialized?

For example:
```
div.style.textDecoration = "none solid";
console.log(div.style.textDecoration); // output is "none"

div.style.textDecoration = "none solid currentcolor";
console.log(div.style.textDecoration); // output is "none"
```
In this example, all keywords, including color, are default value, so we just serialize it as `none`.

Sorry, we land this change into [`text-decoration-valid.html`](https://github.com/web-platform-tests/wpt/blob/master/css/css-text-decor/parsing/text-decoration-valid.html) before discussing with you. Perhaps it's worth to write it down in the spec.


gCS() is a different case. `currentcolor` is always converted into an rgb value, so
```
div.style.textDecoration = "none solid currentcolor";
console.log(getComputedStyle(div).textDecoration);
// output is "rgb(0, 0, 255)", assume currentcolor is blue.
```

Note: I intend to land the serialization for gCS() on `text-decoration` in mozilla [bug 1574222](https://bugzilla.mozilla.org/show_bug.cgi?id=1574222).

**Review by ewilligers — APPROVED** [review]
Do the specs specify where text-decoration-thickness appears in the text-decoration shorthand?

I only see "This property, which is also a sub-property of the text-decoration shorthand, ..."

**Comment by BorisChiou** [comment]
> Do the specs specify where text-decoration-thickness appears in the text-decoration shorthand?
> 
> I only see "This property, which is also a sub-property of the text-decoration shorthand, ..."

I think the answer is no. This is another spec issue, so I need to file one for this. Anyway, I put it in the last one in the text-decoration shorthand for now.

BTW, I found there are some other test cases also need to be updated. I will upload an extra patch for review. Thanks.

**Review by ewilligers — APPROVED** [review]

**Comment by BorisChiou** [comment]
Rebased and try to rerun tests because we hit the Github API issue: https://github.com/web-platform-tests/wpt/issues/18704

**Pull request merged**